blitz_api/views.py

```diff
@@ -1,61 +1,433 @@
+            MAIL_SERVICE["API_KEY"],
+        )
+
+        response_send_mail = email.send_templated_email(
+            email_from=MAIL_SERVICE["EMAIL_FROM"],
+            template_id=MAIL_SERVICE["TEMPLATES"]["FORGOT_PASSWORD"],
+            list_to=[user.email],
+            context={
+                "forgot_password_url": button_url,
+            },
+        )
+
+        if response_send_mail["code"] == "failure":
+            content = {
+                'detail': _("Your token has been created but no email "
+                            "has been sent. Please contact the "
+                            "administration."),
+            }
+            return Response(content, status=status.HTTP_201_CREATED)
+
+        else:
+            return Response(status=status.HTTP_201_CREATED)
+
+
+class ChangePassword(APIView):
+    """
+    post:
+    Get a token and a new password and change the password of
+    the token's owner.
+    """
+    authentication_classes = ()
+    permission_classes = ()
+    serializer_class = serializers.UserSerializer
+
+    def post(self, request):
+        # Valid params
+        serializer = serializers.ChangePasswordSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+
+        token = request.data.get('token')
+        new_password = request.data.get('new_password')
+
+        tokens = ActionToken.objects.filter(
+            key=token,
+            type='password_change',
+            expired=False,
+        )
+
+        # There is only one reference, we will change the user password
+        if len(tokens) == 1:
+            user = tokens[0].user
+            try:
+                password_validation.validate_password(password=new_password)
+            except ValidationError as err:
+                content = {
+                    'detail': err,
+                }
+                return Response(content, status=status.HTTP_400_BAD_REQUEST)
+
+            user.set_password(new_password)
+            user.save()
+
+            # We expire the token used
+            tokens[0].expire()
+
+            # We return the user
+            serializer = serializers.UserSerializer(
+                user,
+                context={'request': request}
+            )
+
+            return Response(serializer.data)
+
+        # There is no reference to this token or multiple identical token
+        # token exists
+        else:
+            error = '{0} is not a valid token.'.format(token)
+
+            return Response(
+                {'detail': error},
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+
+class DomainViewSet(viewsets.ModelViewSet):
+    """
+    retrieve:
+    Return the given domain.
+
+    list:
+    Return a list of all the existing domains.
+
+    create:
+    Create a new domain instance.
+    """
+    serializer_class = serializers.DomainSerializer
+    queryset = Domain.objects.all()
+    permission_classes = (permissions.IsAdminOrReadOnly,)
```

Comment:
I can't filter domain by university and domain are not nested in universities so i can't display (without make a custom filter in frontend) the list of domain the user can use to create his account.

Can you add a filter by university so i can create a little info message to help the user create his account ?

Reply:
The best option would be to add the list of associated domains inside university objects. Will fix that!